### ISSUE TITLE
Callback name inference

### ISSUE THREAD

**Issue by taketwo** [dnouri/skorch#294]
At the moment callback names are either supplied explicitly as first element of a tuple, or are inferred from the callback class name. If two or more callbacks of the same class are used, explicit names have to be used to avoid collision. This leads to the following code (snippet from skorch itself):

```python
    def _default_callbacks(self):
        return [
            ('epoch_timer', EpochTimer()),
            ('train_loss', BatchScoring(
                train_loss_score,
                name='train_loss',
                on_train=True,
                target_extractor=noop,
            )),
            ('valid_loss', BatchScoring(
                valid_loss_score,
                name='valid_loss',
                target_extractor=noop,
            )),
            ('print_log', PrintLog()),
        ]
```

I see some room for improvement here. Note that the strings `train_loss` and `valid_loss` have to be specified twice. First time to define a callback name, second to define a history column name. While technically it's possible to have different names, I doubt it's a realistic use-case. I, for one, always find myself copy-pasting such a callback and then adjusting the name two times, which is annoying. I think it would be very useful if the callback name could be inferred automatically from the history column name.

Second thing to notice is that instead of using automatic inference from the class name (i.e. `EpochTimer` or `PrintLog`), skorch itself opts to explicitly `snake_case` the names. I'd argue that this can be made the default behavior.

To summarize, I propose to update `NeuralNet._yield_callbacks()` such that:
* When callback is an instance and has `name` attribute, it is used.
* When callback name is inferred from class name, it is snake cased.

**Comment by BenjaminBossan** [comment]
You're making good points. It's always a preferable to avoid repetition and be more pythonic. There are, however, some counter-arguments to your points:

Auto-naming: When a user writes their own callback which so happens to have a `name` attribute, this could lead to confusion when their existing code suddenly stops working. Moreover, you could create name conflicts were there currently are none, e.g. when two different callbacks have a `name` attribute with the same value.

Snake-casing: We thought about this one too. But again you have to consider custom callbacks by users. You cannot guarantee that they use camel-case. What do you do with `My_Callback`? You would need a function that unambiguously identifies names that should be snake-cased. And again, the change could break existing code.

In sum, that doesn't mean your proposals are rejected, just that they need some reflection. The current solutions may be ugly at times, but they are also simple and predictable.

PS: For your given example, it could make sense to write a function that returns the `(name, callback)` tuple where you have a single argument for the name that sets both names. In the past, I have written complete sklearn `Pipeline`s (which inspired the callback structure) using custom functions and list comprehensions.

**Comment by taketwo** [comment]
Snake-casing: indeed may be ambiguous and may break existing code, so we can forget about this proposal. (I just spotted this possibility while writing up issue report and decided to include. But I don't really care.)

Auto-naming:

> Moreover, you could create name conflicts were there currently are none, e.g. when two different callbacks have a name attribute with the same value.

These names are used as history column names. If they are same for different callbacks, it means the history is screwed already.

> When a user writes their own callback which so happens to have a name attribute, this could lead to confusion when their existing code suddenly stops working. 

I admit this can happen, though I feel that's a fairly unlikely situation. If we find a way to warn such users about behavior change, I think this may be acceptable.

We can also limit auto-naming to built-in callbacks, or more specifically, those that derive from `ScoringBase.`

**Comment by ottonemo** [comment]
We should also consider dropping the name uniqueness constraint on init and just raise an error if the name cannot be resolved in an unambiguous way. As far as I can see this check is only in place to prevent errors in setparams.

Using a name attribute is OK but it might get intransparent quickly.

**Comment by benjamin-work** [comment]
> These names are used as history column names. If they are same for different callbacks, it means the history is screwed already

Only if the callbacks write to history using the name argument. I would guess that `name` is one of the most popular attribute names.

> We can also limit auto-naming to built-in callbacks, or more specifically, those that derive from ScoringBase.

This makes the behavior even more complicated. Just imagine how many sentences you would need to document the behavior.

> We should also consider dropping the name uniqueness constraint on init and just raise an error if the name cannot be resolved in an unambiguous way

When would you resolve it?

**Comment by ottonemo** [comment]
> When would you resolve it?

In the callback parameter section of `set_params`.

**Comment by benjamin-work** [comment]
> In the callback parameter section of set_params

Is it not better to raise errors as early as possible? Otherwise, it could happen that I lose a lot of time because I only notice late in training that I reused a name.

**Comment by taketwo** [comment]
I guess the point is that this is not an error unless `set_params` is used.

**Comment by benjamin-work** [comment]
> I guess the point is that this is not an error unless set_params is used.

Which still could be bad because more things use `set_params` than might appear at first.

**Comment by taketwo** [comment]
Hm, let me reformulate the issue then. I proposed to infer callback names from the `name` attribute because I saw the pattern. But the matter of fact is, I only care about callback names because skorch is yelling at me if I don't specify them. Thus it'd be great if we find _some_ way of not specifying callback names if you don't need them.

**Comment by benjamin-work** [comment]
How about something analogous to this: http://scikit-learn.org/stable/modules/generated/sklearn.pipeline.make_pipeline.html

**Comment by taketwo** [comment]
It's not clear from the description how it deals with the case of multiple instances of the same class. (Specifically this is my problem, I often have 2-3 different `EpochScoring` callbacks.)

**Comment by benjamin-work** [comment]
```
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
make_pipeline(StandardScaler(), StandardScaler())
# returns
Pipeline(memory=None,
     steps=[('standardscaler-1', StandardScaler(copy=True, with_mean=True, with_std=True)), ('standardscaler-2', StandardScaler(copy=True, with_mean=True, with_std=True))])
```

**Comment by thomasjpfan** [comment]
For reference, here is scikit-learns implementation on estimator naming: https://github.com/scikit-learn/scikit-learn/blob/f1c967883b0ab5efe06a205a404ec6899abc5048/sklearn/pipeline.py#L527-L545

**Comment by benjamin-work** [comment]
@taketwo #301 has been merged. If this solves your issue, please close.

**Comment by taketwo** [comment]
I gave this a try and it does solve my issue. Thanks!
